pkg/client/apiutil/dynamicrestmapper.go

```diff
@@ -0,0 +1,279 @@
+
+	lazy bool
+	// Used for lazy init.
+	initOnce sync.Once
+}
+
+// WithLimiter sets the RESTMapper's underlying limiter to lim.
+func WithLimiter(lim *rate.Limiter) func(*dynamicRESTMapper) error {
+	return func(drm *dynamicRESTMapper) error {
+		drm.limiter = &dynamicLimiter{lim}
+		return nil
+	}
+}
+
+// WithLazyDiscovery prevents the RESTMapper from discovering REST mappings
+// until an API call is made.
+var WithLazyDiscovery = func(drm *dynamicRESTMapper) error {
+	drm.lazy = true
+	return nil
+}
+
+// NewDynamicRESTMapper returns a dynamic RESTMapper for cfg. The dynamic
+// RESTMapper dynamically discovers resource types at runtime. opts
+// configure the RESTMapper.
+func NewDynamicRESTMapper(cfg *rest.Config, opts ...func(*dynamicRESTMapper) error) (meta.RESTMapper, error) {
+	client, err := discovery.NewDiscoveryClientForConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+	drm := &dynamicRESTMapper{
+		client: client,
+		limiter: &dynamicLimiter{
+			rate.NewLimiter(rate.Limit(defaultLimitRate), defaultLimitSize),
+		},
+	}
+	for _, opt := range opts {
+		if err = opt(drm); err != nil {
+			return nil, err
+		}
+	}
+	if !drm.lazy {
+		if err := drm.setStaticMapper(); err != nil {
+			return nil, err
+		}
+	}
+	return drm, nil
+}
+
+var (
+	// defaultLimitRate is the number of RESTMapper API calls allowed
+	// per second assuming the rate of API calls <= defaultLimitRate.
+	defaultLimitRate = 600
+	// defaultLimitSize is the maximum number of simultaneous RESTMapper
+	// API calls allowed.
+	defaultLimitSize = 5
+)
+
+// setStaticMapper sets drm's staticMapper by querying its client, regardless
+// of reload backoff.
+func (drm *dynamicRESTMapper) setStaticMapper() error {
+	groupResources, err := restmapper.GetAPIGroupResources(drm.client)
+	if err != nil {
+		return err
+	}
+	drm.staticMapper = restmapper.NewDiscoveryRESTMapper(groupResources)
+	return nil
+}
+
+// init initializes drm only once if drm is lazy.
+func (drm *dynamicRESTMapper) init() (err error) {
+	drm.initOnce.Do(func() {
+		if drm.lazy {
+			err = drm.setStaticMapper()
+		}
+	})
+	return err
+}
+
+// reload reloads the static RESTMapper, and will return an error only
+// if a rate limit has been hit. reload is thread-safe.
+func (drm *dynamicRESTMapper) reload() error {
+	// limiter is thread-safe.
+	if err := drm.limiter.checkRate(); err != nil {
+		return err
+	}
+	// Lock here so callers can be rate-limited regardless of lock state.
+	drm.mu.Lock()
+	defer drm.mu.Unlock()
+	return drm.setStaticMapper()
+}
+
+// TODO: wrap reload errors on NoKindMatchError with go 1.13 errors.
+
+func (drm *dynamicRESTMapper) KindFor(resource schema.GroupVersionResource) (schema.GroupVersionKind, error) {
+	if err := drm.init(); err != nil {
+		return schema.GroupVersionKind{}, err
+	}
+	drm.mu.RLock()
+	defer drm.mu.RUnlock()
+	gvk, err := drm.staticMapper.KindFor(resource)
+	if xerrors.Is(err, &meta.NoKindMatchError{}) {
+		if rerr := drm.reload(); rerr != nil {
```

Comment:
deadlock here?

RLock --> reload --> Lock

(Go mutexes don't support recursive locking)

Reply:
gotta release the lock here -- we might do duplicate work, but that's probably fine as long as we get the rate limiting (or we could check again after acquiring the write lock or something)

Reply:
do we have tests that exercise this codepath?  Might want those -- would catch things like this

Reply:
You're correct, I misread the docs on how a write lock could be acquired. I'll add unit tests.

Reply:
Perhaps something like:
```go
func (drm *dynamicRESTMapper) reload() error {
	if err := drm.limiter.checkRate(); err != nil {
		return err
	}
	drm.mu.Lock()
	defer drm.mu.Unlock()
	return drm.setStaticMapper()
}

func (drm *dynamicRESTMapper) KindFor(resource schema.GroupVersionResource) (schema.GroupVersionKind, error) {
	if err := drm.init(); err != nil {
		return schema.GroupVersionKind{}, err
	}
	gvk, err := drm.kindFor(resource)
	if xerrors.Is(err, &meta.NoKindMatchError{}) {
		if rerr := drm.reload(); rerr != nil {
			return schema.GroupVersionKind{}, rerr
		}
		gvk, err = drm.kindFor(resource)
	}
	return gvk, err
}

func (drm *dynamicRESTMapper) kindFor(resource schema.GroupVersionResource) (schema.GroupVersionKind, error) {
	drm.mu.RLock()
	defer drm.mu.RUnlock()
	return drm.staticMapper.KindFor(resource)
}
```

Reply:
(in case the thumbs up wasn't obvious, that sounds fine)